### ISSUE TITLE
Hadoop Router Federation On Kerberos  connect failed

### ISSUE THREAD

**Issue by tianyouyangying** [ByConity/ByConity#609]
hadoop3.3.3
ByConity:0.1.2
config:
    <hdfs_nnproxy>hdfs://xxxx</hdfs_nnproxy>
    <hdfs_ha_nameservice>xxxx</hdfs_ha_nameservice>
    <hdfs_user>hadoop</hdfs_user>
    <hdfs3_config>/etc/byconity-server/hdfs3.xml</hdfs3_config>
    <hadoop_kerberos_keytab>/clusteragent/kerberos/keytabs/hadoop.keytab</hadoop_kerberos_keytab>
    <hadoop_kerberos_principal>hadoop/slave133@xxxx.COM</hadoop_kerberos_principal>
    <hadoop_security_authentication>kerberos</hadoop_security_authentication>
bug report:
```
2023.08.16 21:55:17.321560 [ 29042 ] {} <Information> StoragePolicySelector: Storage policy `cnch_default_hdfs` loaded
2023.08.16 21:55:17.321854 [ 29042 ] {} <Information> BackgroundSchedulePool/BgSchPool: Create BackgroundSchedulePool with 6 threads
2023.08.16 21:55:17.322750 [ 29042 ] {} <Information> BackgroundSchedulePool/CNCHTopoPol: Create BackgroundSchedulePool with 4 threads
2023.08.16 21:55:17.324146 [ 29075 ] {} <Information> DiskCacheLoader: Loaded 0 segs from disk /var/lib/byconity-server/
2023.08.16 21:55:17.324250 [ 29075 ] {} <Information> DiskCacheMigrator: Migrated 0 segs from disk /var/lib/byconity-server/
2023.08.16 21:55:17.324287 [ 29075 ] {} <Information> DiskCacheLoader: Loaded 0 segs from disk /var/lib/byconity-server/server_local_disk/data/0/
2023.08.16 21:55:17.324437 [ 29075 ] {} <Information> DiskCacheMigrator: Migrated 0 segs from disk /var/lib/byconity-server/server_local_disk/data/0/
2023.08.16 21:55:17.324475 [ 29042 ] {} <Information> Application: Init cnch storage cache.
2023.08.16 21:55:17.324500 [ 29042 ] {} <Error> CnchServerManager: There is no zookeeper, skip start background task for serverManager
2023.08.16 21:55:17.324815 [ 29070 ] {} <Debug> ThreadPool: clear thread for finalize : [29077, 29075, ]
2023.08.16 21:55:17.325021 [ 29070 ] {} <Information> DiskCacheLRU: load thread takes 2 ms
2023.08.16 21:55:17.426192 [ 29042 ] {} <Debug> CnchServerManager: Successfully set leader status.
2023.08.16 21:55:17.426249 [ 29042 ] {} <Debug> CnchServerManager: Current node 192.168.5.133:8124 become leader
2023.08.16 21:55:17.426261 [ 29042 ] {} <Information> Application: Loading metadata from /var/lib/byconity-server/
2023.08.16 21:55:17.426484 [ 29076 ] {} <Debug> TSOClient[{TSO-0 slave133  rpc/9910}]: Initialized rpc channel to server with address: slave133:9910
2023.08.16 21:55:17.427888 [ 29042 ] {} <Information> DatabaseAtomic (system): Total 0 tables and 0 dictionaries.
2023.08.16 21:55:17.427905 [ 29042 ] {} <Information> DatabaseAtomic (system): Starting up tables.
2023.08.16 21:55:17.428876 [ 29076 ] {} <Debug> CnchServerManager: Removing expired topology : {expiration: 1692212097804, [slave133]}
2023.08.16 21:55:17.428905 [ 29076 ] {} <Debug> CnchServerManager: Add new topology {expiration: 1692212129404, [slave133]}
2023.08.16 21:55:17.435236 [ 29042 ] {} <Information> DatabaseAtomic (default): Total 0 tables and 0 dictionaries.
2023.08.16 21:55:17.435254 [ 29042 ] {} <Information> DatabaseAtomic (default): Starting up tables.
2023.08.16 21:55:17.435283 [ 29042 ] {} <Information> DatabaseCatalog: Found 0 partially dropped tables. Will load them and retry removal.
2023.08.16 21:55:17.436955 [ 29042 ] {} <Debug> SystemLog (system.server_part_log): Creating new table system.server_part_log for ServerPartLog
2023.08.16 21:55:17.439831 [ 29042 ] {} <Debug> StorageFactory:  engine name MergeTree
2023.08.16 21:55:17.440408 [ 29042 ] {} <Debug> system.server_part_log (ea43fa9c-c9de-411e-80e9-3a77e34ef876): Create metastore directory /var/lib/byconity-server/metastore/store/ea4/ea43fa9c-c9de-411e-80e9-3a77e34ef876/ for table system.server_part_log (ea43fa9c-c9de-411e-80e9-3a77e34ef876)
2023.08.16 21:55:17.515328 [ 29042 ] {} <Debug> system.server_part_log (ea43fa9c-c9de-411e-80e9-3a77e34ef876): Start loading data parts from filesystem.
2023.08.16 21:55:17.515841 [ 29042 ] {} <Debug> HDFSClient: Running KerberosInit
2023.08.16 21:55:17.518231 [ 29042 ] {} <Debug> HDFSClient: Finished KerberosInit
2023.08.16 21:55:17.828882 [ 29074 ] {} <Warning> CnchTopologyMaster: Reset part and table cache because of topology change
2023.08.16 21:55:17.828946 [ 29074 ] {} <Debug> PartCacheManager::reset: Resetting part cache manager.
2023.08.16 21:55:18.803908 [ 29042 ] {} <Debug> HDFS3: Failed to setup RPC connection to "slave01:8888" caused by:
RpcChannel.cpp: 868: GSS initiate failed
GSS initiate failed
2023.08.16 21:55:18.803939 [ 29042 ] {} <Debug> HDFS3: Retrying connect to server: "slave01:8888". Already tried 1 time(s)
2023.08.16 21:55:22.097720 [ 29042 ] {} <Debug> HDFS3: Failed to setup RPC connection to "slave01:8888" caused by:
RpcChannel.cpp: 868: GSS initiate failed
GSS initiate failed
2023.08.16 21:55:22.097750 [ 29042 ] {} <Debug> HDFS3: Retrying connect to server: "slave01:8888". Already tried 2 time(s)
2023.08.16 21:55:27.428992 [ 29073 ] {} <Debug> CnchSystemLogs: Initialise CNCH System log on try: 1
2023.08.16 21:55:27.429036 [ 29073 ] {} <Information> CnchSystemLogs: Initializing CNCH System log on server
2023.08.16 21:55:28.396110 [ 29042 ] {} <Debug> HDFS3: Failed to setup RPC connection to "slave01:8888" caused by:
RpcChannel.cpp: 868: GSS initiate failed
GSS initiate failed
2023.08.16 21:55:28.396142 [ 29042 ] {} <Debug> HDFS3: Retrying connect to server: "slave01:8888". Already tried 3 time(s)
2023.08.16 21:55:30.706967 [ 29042 ] {} <Debug> HDFS3: Failed to setup RPC connection to "slave01:8888" caused by:
RpcChannel.cpp: 868: GSS initiate failed
GSS initiate failed
2023.08.16 21:55:30.706999 [ 29042 ] {} <Debug> HDFS3: Retrying connect to server: "slave01:8888". Already tried 4 time(s)
2023.08.16 21:55:33.238396 [ 29042 ] {} <Debug> HDFS3: Failed to setup RPC connection to "slave01:8888" caused by:
RpcChannel.cpp: 868: GSS initiate failed
GSS initiate failed
2023.08.16 21:55:33.238429 [ 29042 ] {} <Debug> HDFS3: Retrying connect to server: "slave01:8888". Already tried 5 time(s)
2023.08.16 21:55:39.509417 [ 29042 ] {} <Debug> HDFS3: Failed to setup RPC connection to "slave01:8888" caused by:
RpcChannel.cpp: 868: GSS initiate failed
GSS initiate failed
2023.08.16 21:55:39.509445 [ 29042 ] {} <Debug> HDFS3: Retrying connect to server: "slave01:8888". Already tried 6 time(s)
2023.08.16 21:55:45.793000 [ 29042 ] {} <Debug> HDFS3: Failed to setup RPC connection to "slave01:8888" caused by:
RpcChannel.cpp: 868: GSS initiate failed
GSS initiate failed
2023.08.16 21:55:45.793031 [ 29042 ] {} <Debug> HDFS3: Retrying connect to server: "slave01:8888". Already tried 7 time(s)
2023.08.16 21:55:50.121127 [ 29042 ] {} <Debug> HDFS3: Failed to setup RPC connection to "slave01:8888" caused by:
RpcChannel.cpp: 868: GSS initiate failed
GSS initiate failed
2023.08.16 21:55:50.121160 [ 29042 ] {} <Debug> HDFS3: Retrying connect to server: "slave01:8888". Already tried 8 time(s)
2023.08.16 21:55:55.394794 [ 29042 ] {} <Debug> HDFS3: Failed to setup RPC connection to "slave01:8888" caused by:
RpcChannel.cpp: 868: GSS initiate failed
GSS initiate failed
2023.08.16 21:55:55.394825 [ 29042 ] {} <Debug> HDFS3: Retrying connect to server: "slave01:8888". Already tried 9 time(s)
2023.08.16 21:55:58.676951 [ 29042 ] {} <Debug> HDFS3: Failed to setup RPC connection to "slave01:8888" caused by:
RpcChannel.cpp: 868: GSS initiate failed
GSS initiate failed
2023.08.16 21:56:00.679672 [ 29042 ] {} <Debug> HDFS3: Method: getFileInfo Request: src: "/user/clickhouse/store/ea4/ea43fa9c-c9de-411e-80e9-3a77e34ef876" Response: 
2023.08.16 21:56:00.697637 [ 29042 ] {} <Error> Application: Caught exception while loading metadata: Poco::Exception. Code: 1000, e.code() = 13, e.displayText() = Access to file denied: GSS initiate failed, Stack trace (when copying this message, always include the lines below):

0. /build/build_docker/../contrib/libcxx/include/exception:133: Poco::FileAccessDeniedException::FileAccessDeniedException(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, int) @ 0x19eaac92 in /usr/bin/clickhouse
1. /build/build_docker/../src/Storages/HDFS/HDFSFileSystem.cpp:656: DB::HDFSFileSystem::handleError(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) const @ 0x13d94e57 in /usr/bin/clickhouse
2. /build/build_docker/../src/Storages/HDFS/HDFSFileSystem.cpp:0: DB::HDFSFileSystem::exists(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) const @ 0x13d96183 in /usr/bin/clickhouse
3. /build/build_docker/../src/Disks/HDFS/DiskByteHDFS.cpp:0: DB::DiskByteHDFS::exists(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) const @ 0x13dae0cf in /usr/bin/clickhouse
4. /build/build_docker/../src/Storages/MergeTree/MergeTreeData.cpp:550: DB::MergeTreeData::loadDataParts(bool) @ 0x1514dd8b in /usr/bin/clickhouse
5. /build/build_docker/../src/Storages/StorageMergeTree.cpp:113: DB::StorageMergeTree::StorageMergeTree(DB::StorageID const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, DB::StorageInMemoryMetadata const&, bool, std::__1::shared_ptr<DB::Context>, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, DB::MergeTreeMetaBase::MergingParams const&, std::__1::unique_ptr<DB::MergeTreeSettings, std::__1::default_delete<DB::MergeTreeSettings> >, bool) @ 0x14e939d6 in /usr/bin/clickhouse
6. /build/build_docker/../base/common/../common/shared_ptr_helper.h:0: DB::create(DB::StorageFactory::Arguments const&) @ 0x1534bebe in /usr/bin/clickhouse
7. /build/build_docker/../src/Storages/StorageFactory.cpp:241: DB::StorageFactory::get(DB::ASTCreateQuery const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::shared_ptr<DB::Context>, std::__1::shared_ptr<DB::Context>, DB::ColumnsDescription const&, DB::ConstraintsDescription const&, bool) const @ 0x14e104be in /usr/bin/clickhouse
8. /build/build_docker/../contrib/libcxx/include/type_traits:3935: DB::InterpreterCreateQuery::doCreateTable(DB::ASTCreateQuery&, DB::InterpreterCreateQuery::TableProperties const&) @ 0x1474f057 in /usr/bin/clickhouse
9. /build/build_docker/../src/Interpreters/InterpreterCreateQuery.cpp:1118: DB::InterpreterCreateQuery::createTable(DB::ASTCreateQuery&) @ 0x1474b406 in /usr/bin/clickhouse
10. /build/build_docker/../src/Interpreters/InterpreterCreateQuery.cpp:0: DB::InterpreterCreateQuery::execute() @ 0x14750cd9 in /usr/bin/clickhouse
11. /build/build_docker/../src/Interpreters/SystemLog.h:0: DB::SystemLog<DB::ServerPartLogElement>::prepareTable() @ 0x1494cba4 in /usr/bin/clickhouse
12. /build/build_docker/../contrib/libcxx/include/memory:3211: DB::Server::main(std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > > const&) @ 0xb795509 in /usr/bin/clickhouse
13. /build/build_docker/../contrib/poco/Util/src/Application.cpp:0: Poco::Util::Application::run() @ 0x19e2c64a in /usr/bin/clickhouse
14. /build/build_docker/../programs/server/Server.cpp:513: DB::Server::run() @ 0xb789944 in /usr/bin/clickhouse
15. /build/build_docker/../contrib/poco/Util/src/ServerApplication.cpp:612: Poco::Util::ServerApplication::run(int, char**) @ 0x19e3e4bb in /usr/bin/clickhouse
16. /build/build_docker/../programs/server/Server.cpp:0: mainEntryClickHouseServer(int, char**) @ 0xb787c22 in /usr/bin/clickhouse
17. /build/build_docker/../programs/main.cpp:0: main @ 0xb6fda68 in /usr/bin/clickhouse
18. __libc_start_main @ 0x22555 in /usr/lib64/libc-2.17.so
19. _start @ 0xb6fd26a in /usr/bin/clickhouse
 (version 21.8.7.1)
2023.08.16 21:56:00.697755 [ 29042 ] {} <Information> Application: Shutting down storages.
2023.08.16 21:56:00.697764 [ 29042 ] {} <Information> Context: Shutdown disk default
2023.08.16 21:56:00.697769 [ 29042 ] {} <Information> Context: Shutdown disk server_hdfs_disk0
2023.08.16 21:56:00.697773 [ 29042 ] {} <Information> Context: Shutdown disk server_local_disk1
2023.08.16 22:05:17.324414 [ 29072 ] {} <Information> TransactionCoordinator: Background txn scan task starts..
```

**Comment by white-cc** [comment]
You can specify the parameter `-DENABLE_EXAMPLES=1` with cmake
and `ninja kerberos_init` and `ninja hdfs_krb5`
1. **kerberos_init**
You can test the validity of kerberos_keytab and kerberos_principal (execute kinit)
Command:` ./kerberos_init keytab principal [cache]`
Code location: src/Access/examples/Kerberos_init.cpp
2. **hdfs_krb5**
You can test the success of kerberos authentication to access hdfs (list the hdfs directory)
Command: `./hdfs_krb5 keytab principal cache_path hdfs_path`
Code location: src/Storages/HDFS/examples/Hdfs_krb5.cpp

For krb5 authentication problems, most of them are configuration krb5 problems, which can set
`export KRB5_TRACE=/dev/stdout` to debug krb5 problems

**Comment by white-cc** [comment]
https://web.mit.edu/kerberos/krb5-1.12/doc/admin/troubleshoot.html

**Comment by tianyouyangying** [comment]
Thanks
Problem solved, but there are new problems
new error:
```
2023.08.17 10:33:16.771701 [ 21704 ] {} <Warning> Application: Calculated checksum of the binary: C65EC31CCF304B775A55DCD79817035B. There is no information about the reference checksum.
2023.08.17 10:33:16.771892 [ 21704 ] {} <Debug> Application: rlimit on number of file descriptors is 500000
2023.08.17 10:33:16.771915 [ 21704 ] {} <Debug> Application: Initializing DateLUT.
2023.08.17 10:33:16.771936 [ 21704 ] {} <Debug> Application: Setting up /var/lib/byconity-server/tmp_data/ to store temporary data in it
2023.08.17 10:33:16.772159 [ 21704 ] {} <Debug> Application: Configuration parameter 'interserver_http_host' doesn't exist or exists and empty. Will use 'localhost' as replica host.
2023.08.17 10:33:16.772182 [ 21704 ] {} <Debug> Application: Initiailizing interserver credentials.
2023.08.17 10:33:16.773411 [ 21704 ] {} <Debug> ConfigReloader: Loading config '/etc/byconity-server/byconity-server.xml'
2023.08.17 10:33:16.774308 [ 21704 ] {} <Debug> ConfigReloader: Loaded config '/etc/byconity-server/byconity-server.xml', performing update on configuration
2023.08.17 10:33:16.774990 [ 21704 ] {} <Information> Application: Setting max_server_memory_usage was set to 14.10 GiB (15.67 GiB available * 0.90 max_server_memory_usage_to_ram_ratio)
2023.08.17 10:33:16.775250 [ 21704 ] {} <Debug> ConfigReloader: Loaded config '/etc/byconity-server/byconity-server.xml', performed update on configuration
2023.08.17 10:33:16.777196 [ 21704 ] {} <Debug> Access(user directories): Added users.xml access storage 'users.xml', path: 
2023.08.17 10:33:16.777240 [ 21704 ] {} <Debug> void DB::Context::initResourceGroupManager(const DB::Context::ConfigurationPtr &): Skip initialize resource group
2023.08.17 10:33:16.778424 [ 21704 ] {} <Debug> ConfigReloader: Loading config '/etc/byconity-server/users.xml'
2023.08.17 10:33:16.779187 [ 21704 ] {} <Debug> ConfigReloader: Loaded config '/etc/byconity-server/users.xml', performing update on configuration
2023.08.17 10:33:16.779540 [ 21704 ] {} <Debug> ConfigReloader: Loaded config '/etc/byconity-server/users.xml', performed update on configuration
2023.08.17 10:33:16.779836 [ 21704 ] {} <Debug> Access(user directories): Added users.xml access storage 'users.xml', path: /etc/byconity-server/users.xml
2023.08.17 10:33:16.779871 [ 21704 ] {} <Information> Application: Uncompressed cache size was lowered to 7.83 GiB because the system has low amount of memory
2023.08.17 10:33:16.780041 [ 21704 ] {} <Information> KeyIndexFileCache: Initializing KeyIndexFileCache...
2023.08.17 10:33:16.780083 [ 21704 ] {} <Information> KeyIndexFileCache: Initialized KeyIndexFileCache with 0 entries in 0 ms
2023.08.17 10:33:16.781889 [ 21704 ] {} <Error> Application: DB::Exception: create builder on uninitialized HDFSConecitonParams SQLSTATE: HY000
2023.08.17 10:33:16.781931 [ 21704 ] {} <Information> Application: shutting down
2023.08.17 10:33:16.781936 [ 21704 ] {} <Debug> Application: Uninitializing subsystem: Logging Subsystem
2023.08.17 10:33:16.782039 [ 21714 ] {} <Information> BaseDaemon: Stop SignalListener thread

```

**Comment by white-cc** [comment]
Can you paste the relevant log about the NameNode & DataNode? 
And make sure hdfs3.xml settings is consistent with your hdfs config settings.
